design_docs/rsyslog_new_input_output_roles_proposal.md

```diff
@@ -0,0 +1,255 @@
+
+For the comparison, here is the previous style, in which logs_collections (input roles) used to belong to logging_outputs (output roles) to specify the relationship between the input and the output roles.
+
+**Previous style**
+```
+logging_outputs:
+  - name: forward-output
+    type: forwards
+    rsyslog_forwards_actions:
+      - name: to-remote
+        protocol: tcp
+        target: remote_host_name.remote_domain_name
+        port: 514
+    logs_collections:
+      - name: basic-input
+        type: basics
+  - name: file-output
+    type: files
+    logs_collections:
+      - name: file-input
+        type: files
+```
+In the above example, the inputs from imjournal (basic-input) are forwarded to the remote host (forward-output); the inputs from imfile (file-input) are stored in the local file (file-output).
+
+With this style, it is not straightforward to implement one input to send multiple outputs in the logging roles.
+Moreover, to configure the input from `imjournal` to send both the remote host and the local file,
+the basics input would be placed under both forwards and files outputs as follows.
+It could give an impression basic-input0 and basic-input1 could be configured independently, but it is not.
+There is only one `imjournal` input and most likely the first one is wiped out.
+If possible, we should avoid this type of confusions in the designing phase.
+
+**Previous style example with uncertainty**
+```
+logging_outputs:
+  - name: forward-output
+    type: forwards
+    rsyslog_forwards_actions:
+      - name: to-remote
+        protocol: tcp
+        target: remote_host_name.remote_domain_name
+        port: 514
+    logs_collections:
+      - name: basic-input0
+        type: basics
+        rsyslog_imjournal_persist_state_interval: 1000
+        rsyslog_imjournal_ratelimit_interval: 10
+        rsyslog_imjournal_ratelimit_burst: 100000
+  - name: file-output
+    type: files
+    logs_collections:
+      - name: basic-input1
+        type: basics
+        rsyslog_imjournal_persist_state_interval: 3000
+        rsyslog_imjournal_ratelimit_interval: 30
+        rsyslog_imjournal_ratelimit_burst: 300000
+```
+
+### logging_outputs, logs_collections and rsyslog_use_OUTPUT_ruleset
+This is a typical example of the configuration.
+In logging_outputs, the to-be-configured output roles are listed.
+In logs_collections, the to-be-configure input roles are listed.
+Each input role has `rsyslog_use_OUTPUT_ruleset: true|false` to specify which output the log messages are sent to.
+```
+logging_outputs:
+  - name: forward-output
+    type: forwards
+    rsyslog_forwards_actions:
+      - name: to-remote
+        protocol: tcp
+        target: remote_host_name.remote_domain_name
+        port: 514
+  - name: file-output
+    type: files
+logs_collections:
+  - name: basic-input
+    type: basics
+    rsyslog_use_files_ruleset: true
```

Comment:
Isn't it better to refer to outputs by their name? I.e. if file output is defined as `name: file-output`, refer to it by the name `file-output`, rather than using a boolean. This would make sense especially if one can specify multiple outputs of the same type, because then how will you distinguish between them otherwise?

Reply:
The purpose of this implementation is to call `call` statement as simple as this.
https://github.com/nhosoi/logging/blob/combination/roles/rsyslog/roles/input_roles/files/defaults/main.yml#L66-L71

We need to have an ability to deliver the log messages to one particular output as well as multiple ones.  That being said, are you suggesting something like this?  (please note that instead of `name`, i'm putting `type` in the array since only one instance is available per one output role as described in https://github.com/linux-system-roles/logging/pull/100#discussion_r401131546.) If this looks ok, I will update this patch.
```
logs_collections:
  - name: basic-input
    type: basics
    rsyslog_outputs: ['files', 'forwards']
```

Reply:
yes, something like that. But I still think that putting `name` instead of `type` in the array makes more sense, because `name` is supposed to be the unique identifier. The role can resolve `name` to type internally. If the role starts supporting multiple output declaration for each type, it will be easy to support without changing the interface. (And if you think that this will never be supported, why having separate `name` and `type` at all?)

Reply:
> yes, something like that. But I still think that putting `name` instead of `type` in the array makes more sense, because `name` is supposed to be the unique identifier. The role can resolve `name` to type internally. If the role starts supporting multiple output declaration for each type, it will be easy to support without changing the interface. (And if you think that this will never be supported, why having separate `name` and `type` at all?)

Hmmm, I thought the value of `name` could be anything that the customers name it in their configuration, on the other hand, the `type` could be restricted in the given types.  But it might have been a wrong impression.

Here's the summary:
1) only one instance per one type of input and output role is allowed.  (the rests are ignored.)
2) name of the input and output role has to be in [ 'basics', 'files', 'forwards' ] (note: 'files' is used for the input, imfile, as well aw for the output, omfile).

1 is a fact, if 2 is acceptable, I can eliminate the type subvariable and use name instead.

Then, the simplest configuration (imjournal + omfile) would look like this:
```
        logging_inputs:
          - name: basics
```
@sradco, @richm, I'd like to have your thoughts, too.  Thanks!

Reply:
I would keep `type`, I think it expresses the meaning better. Would it be possible to have `name` default to `type` if omitted? What would customers use `name` for, is it reflected in any way in the resulting configuration? (I would think that it is used to construct the resulting configuration snippet name.)

Reply:
I will eliminate `name` and keep `type`.
```
        logging_inputs:
          - type: basics
```

Reply:
Please, what is the answer to
> What would customers use `name` for, is it reflected in any way in the resulting configuration?

Reply:
Could `name` be used for the `name` field of the `input()`?  If so, might be useful for `impstats` or flow control: https://www.rsyslog.com/doc/v8-stable/configuration/modules/imtcp.html#name

Reply:
> Could `name` be used for the `name` field of the `input()`? If so, might be useful for `impstats` or flow control: https://www.rsyslog.com/doc/v8-stable/configuration/modules/imtcp.html#name

Yep, it is.  Also, in the action.  I'm not using it in the files action since it's in the old format.  I guess I should update them.  But breaking the elasticsearch use case is more destructive...  (please see https://github.com/linux-system-roles/logging/pull/100#discussion_r403132931)  I'm reworking on it.